src/github.com/matrix-org/dendrite/common/transactions/transactions_test.go

```diff
@@ -0,0 +1,75 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transactions
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/matrix-org/util"
+)
+
+type fakeType struct {
+	ID string `json:"ID"`
+}
+
+var (
+	fakeTxnID    = "aRandomTxnID"
+	fakeResponse = &util.JSONResponse{Code: http.StatusOK, JSON: fakeType{ID: "0"}}
+)
+
+// TestCache creates a New Cache and tests AddTransaction & FetchTransaction
+func TestCache(t *testing.T) {
+
+	fakeTxnCache := New()
+	fakeTxnCache.AddTransaction(fakeTxnID, fakeResponse)
+
+	// Add entries for noise.
+	for i := 1; i <= 100; i++ {
+		fakeTxnCache.AddTransaction(
+			fakeTxnID+string(i),
+			&util.JSONResponse{Code: http.StatusOK, JSON: fakeType{ID: string(i)}},
+		)
+	}
+
+	testResponse, err := fakeTxnCache.FetchTransaction(fakeTxnID)
+	if err != nil {
+		t.Error("Failed to retrieve entry for txnID: ", fakeTxnID)
+	} else if testResponse.JSON != fakeResponse.JSON {
+		t.Error("Incorrect fetched response. Expected: ", fakeResponse.JSON, " got: ", testResponse.JSON)
+	}
+}
+
+// TestConcurentAccess provides some guarantee against corruption.
+func TestConcurentAccess(t *testing.T) {
+	fakeTxnCache := New()
+	// Signal that this test should run in parallel
+	t.Parallel()
+	// Add entries concurrently to test concurrent access.
+	for i := 1; i <= 1000; i++ {
+		go check(t, fakeTxnCache, i)
+	}
+}
```

Comment:
Honestly, I'm not sure this test is really testing anything much; if there is a concurrency bug then this is unlikely to pick it up. The concurrency model here is to simply use `RWLock`, so I think its safe to simply not have a unit test and to manually make sure we lock in the appropriate methods.

(If the cache was more complicated, then we'd probably want to add hooks into the cache to let us write deterministic tests.)